Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python3
+
+from http.server import SimpleHTTPRequestHandler, HTTPServer
+import argparse
+import logging
+import os
+import sys
+import socket
+import threading
+# Since we execute this script directly as a subprocess, we need to ensure
+# that Tools/Scripts is in sys.path for the next imports to work correctly.
+script_dir = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.abspath(__file__)), '../../../..'))
+if script_dir not in sys.path:
+    sys.path.insert(0, script_dir)
+
+
+_log = logging.getLogger(__name__)
+
+
+class Handler(SimpleHTTPRequestHandler):
+    web_root = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs, directory=self.web_root)
+
+    def log_message(self, format, *args):
+        _log.info(format % args)
+
+    def do_GET(self):
+        if self.path.startswith('/shutdown'):
+            self.send_response(200)
+            self.end_headers()
+            try:
+                shutdown_thread = threading.Thread(target=self.server.shutdown)
+                shutdown_thread.start()
+                sys.exit(0)
+            except Exception as e:
+                _log.error('Error shutting down HTTP server: %s' % e)
+                sys.exit(1)
+
+        else:
+            super().do_GET()
+
+    def do_POST(self):
+        if self.path.startswith('/report'):
+            length = int(self.headers['content-length'])
+            try:
+                content = self.rfile.read(length)
+                sys.stdout.write(content.decode('utf-8'))
+                sys.stdout.flush()
+                self.send_response(200)
+            except Exception as e:
+                _log.error('Error reading and writing report: %s' % e)
+                self.send_response(500)
+            self.end_headers()
+            try:
+                shutdown_thread = threading.Thread(target=self.server.shutdown)
+                shutdown_thread.start()
+                sys.exit(0)
+            except Exception as e:
+                _log.error('Error shutting down HTTP server: %s' % e)
+                sys.exit(1)
+        else:
+            super().do_POST()
+
+
+def find_free_port():
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.bind(('', 0))
+    port = s.getsockname()[1]
+    s.close()
+    return port
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description='python base_http_server.py web_root')
+    parser.add_argument('web_root')
+    parser.add_argument('--port', type=int, default=0)
+    parser.add_argument('--interface', default='')
+    parser.add_argument('--log-path', default='/tmp/run-benchmark-http.log')
+    args = parser.parse_args()
+
+    logging.basicConfig(datefmt='%Y-%m-%d %H:%M:%S',
+                        format='%(asctime)s [%(levelname)s]: %(message)s', filename=args.log_path, level=logging.INFO)
+    port = args.port
+    if port == 0:
+        try:
+            port = find_free_port()
```

Comment:
This is technically subject to race condition.
By digging the source code a bit (https://github.com/python/cpython/blob/3.11/Lib/socketserver.py#L472C26-L472C45 ), I think `HTTPServer((args.interface, port), Handler)` will do exactly what we want: when port == 0, find a free port; when port != 0, try to bind the specified port.

Reply:
If its subject to race condition then I'll let socketserver handle finding a free port